sdk/servicebus/Microsoft.Azure.ServiceBus/src/Core/MessageReceiver.cs

```diff
@@ -1295,6 +1344,8 @@ protected virtual async Task<DateTime> OnRenewLockAsync(string lockToken)
                     {
                         this.receivePumpCancellationTokenSource.Cancel();
                         this.receivePumpCancellationTokenSource.Dispose();
+                        //this.runningTaskCancellationTokenSource.Cancel();
```

Comment:
Not sure if we should cancel and/or dispose the runningTaskCancellationTokenSource here since if unregister was called before and there are some pending tasks to be done, if the next register call run into this exception, we would terminate those pending tasks if the cancellation token is fired.

Reply:
If you have a lock around the entire implementation of UnregisterMessageHandler including the while loop waiting for in-flight executions to complete, this situation should not arise.  i.e. a call to RegisterMessageHandler while UnregisterMessageHandler is executing will block until UnregisterMessageHandler completes or has an exception (lock released) after which it is safe to cancel and dispose the runningTaskCancellationTokenSource.

What is the reason that you don't want to hold the lock while draining in-flight executions?

Reply:
We have await statement inside the while loop so we cannot put that while loop inside the lock since it will potentially cause deadlock.

Reply:
I see.  Cancelling and disposing is ok here.